src/util/object_pool.js

```diff
@@ -0,0 +1,49 @@
+/* Copyright (C) 2016 NooBaa */
+'use strict';
+
+const { make_array } = require('./js_utils');
+
+const alloc_empty_obj = () => ({});
+const noop = () => { /* noop */ };
+
+class ObjectPool {
+    constructor(capacity, allocator, initializer, empty_value = null) {
+        if (!capacity) {
+            console.warn('ObjectPool - Invalid operation, cannot create a pool without capacity');
+        }
+
+        this._capacity = capacity;
+        this._initalize = initializer || noop;
+        this._empty_value = empty_value;
+        this._next = 0;
+
+        const allocate = allocator || alloc_empty_obj;
+        this._items = make_array(capacity, allocate);
+    }
+
+    alloc() {
+        if (this._next === this._capacity) {
+            console.warn('ObjectPool - Could not allocate, pool is empty');
+            return this._empty_value;
+        }
+
+        const item = this._items[this._next];
+        this._items[this._next] = this._empty_value;
+        this._next += 1;
+
+        return item;
+    }
+
+    release(item) {
+        if (this._next === 0) {
+            console.warn('ObjectPool - Could not release, pool is at full capacity');
```

Comment:
Why not grow it to double the previous capacity?

Reply:
In the general case, you would like to grow the pool in case you cannot allocate any more. Here we are trying to return something to the pool but everything is already in the pool which means something was returned to the pool that was not created from the pool.

Regarding the `alloc()` case, I wanted a stricter pool that enforces a fixed size no preserve the no allocation rule. 
If we will use this for other use cases we can add code to decide what behavior we want upon pool creation (Fixed, Resize, Shrink, etc.)

Reply:
I have added a resize_policy as an option the `ObjectPool` constructor, currently, I support the following: `NO_RESIZE`, `GROW_WHEN_EMPTY` (default), `GORW_AND_SHRINK`
I believe this will suffice for most use cases.
I also externalize the capacity and availability of the pool